staging/src/k8s.io/api/flowcontrol/v1alpha1/types.go

```diff
@@ -0,0 +1,232 @@
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// FlowSchemaList is a list of FlowSchema objects.
+type FlowSchemaList struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard list metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ListMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	// Items is a list of flow-schemas.
+	Items []FlowSchema `json:"items" protobuf:"bytes,2,rep,name=items"`
+}
+
+// FlowSchemaSpec describes how the flow-schema's specification looks like.
+type FlowSchemaSpec struct {
+	// PriorityLevelConfigurationRef can reference a PriorityLevelConfiguration in the cluster. If the reference cannot
+	// be resolved, the flow-schema will ignored and marked as invalid in its status.
+	PriorityLevelConfiguration PriorityLevelConfigurationReference `json:"priorityLevelConfiguration,omitempty" protobuf:"bytes,1,opt,name=priorityLevelConfiguration"`
+	// MatchingPrecedence is used to choose among the FlowSchemas that match a given request.  The chosen
+	// FlowSchema is among those with the numerically lowest (which we take to be logically highest)
+	// MatchingPrecedence.  Each MatchingPrecedence value must be non-negative.
+	MatchingPrecedence int32 `json:"matchingPrecedence,omitempty" protobuf:"varint,2,opt,name=matchingPrecedence"`
+	// DistinguisherMethod defines how to classify requests into certain flows. If not empty, the flow's identifier is computed
+	// according to specification of the distinguisher.
+	// +optional
+	DistinguisherMethod FlowDistinguisherMethod `json:"distinguisherMethod,omitempty" protobuf:"bytes,3,opt,name=distinguisherMethod"`
+	// Rules holds all the RuleWithUser for this FlowSchema.
+	Rules []PolicyRuleWithSubject `json:"rules,omitempty" protobuf:"bytes,4,rep,name=rules"`
+}
+
+// FlowDistinguisherMethodType is the type of flow distinguisher method
+type FlowDistinguisherMethodType string
+
+// FlowDistinguisherMethod specifies the method of a flow distinguisher.
+type FlowDistinguisherMethod struct {
+	// Type is the type of flow distinguisher method
+	Type FlowDistinguisherMethodType `json:"type,omitempty" protobuf:"varint,1,opt,name=type"`
+}
+
+// PriorityLevelConfigurationReference contains information that points to the request-priority being used.
+type PriorityLevelConfigurationReference struct {
+	// Name is the name of resource being referenced
+	Name string `json:"name,omitempty" protobuf:"varint,1,opt,name=name"`
+}
+
+// PolicyRuleWithSubject prescribes a test that applies to a request to an apiserver. The test considers the subject
+// making the request, the verb being requested, and the resource to be acted upon.
+type PolicyRuleWithSubject struct {
+	// Subject is the normal user, serviceaccount, or group that this rule cares about.
+	// +optional
+	Subject rbacv1.Subject `json:"subject,omitempty" protobuf:"bytes,1,opt,name=subject"`
```

Comment:
Sorry for not noticing this earlier. But the referenced type is for the purpose of describing a subject; however, the task of this api object is to match some subject(s). I do not think it is a good idea to require a rule for every subject.

Maybe follow the style of policy rule below?

```
// `users` is a list of users which this rule will match. 
Users []string

// `groups` is a list of groups which this rule will match.
Groups []string

// If `allSubjects` is true, this rule matches all subjects.
AllSubjects bool
```

Reply:
Alternatively I can imagine include / exclude lists. But I think this is simpler. Folks wanting exclude lists can make a different rule that takes precedence.

Reply:
This is why I started with a predicate-based approach.  It smoothly supports all the things you mentioned.

Another non-predicate approach would be to replace `Subject rbacv1.Subject` with `Subjects []rbacv1.Subject` and define a `rbacv1.Subject` value that means "all".

The `rbacv1.Subject` struct is surprisingly complex.  Why does it have a `Namespace` field?  I thought that every service account is a user and includes a namespace in the username (which goes in the `Name` field).  Why does it have an `APIGroup` field?  That seems unnecessary.

The `rbacv1.Subject` struct follows a divergent pattern for unions.  We do them differently, for example, for volumes.

Either there is a good reason for the complexity in the `rbacv1.Subject` struct and we should keep using it here, or we might want to define our own alternative that is simpler and follows the usual pattern for unions.  Perhaps something like this:
```
// Subject identifies a set of users.  Exactly one field should be non-zero.
type Subject struct {
    // UserName matches one user, the one with the given name, or all users if the given name is UserNameAll.
    UserName string `json:"userName,omitempty" protobuf:"bytes,1,opt,name=userName"`

    // Group matches all users that belong to the named group.
    Group string `json:"group,omitempty" protobuf:"bytes,2,opt,name=group"`
}

// UserNameAll matches all usernames
const UserNameAll = "*"
```

In this case a `PolicyRuleWithSubject` could have a field `Subjects []Subject`.

Reply:
how do you think about:

```go
// Operator has to be one of In and NotIn
Operator ListOperator 
// Subjects is a list of subject
Subjects []rbacv1.Subject
```

it's basically the same as the include/exclude lists as daniel say. what's more, it's consistent w/ the api definitions from the rbac world.
```yaml
rules:
- operator: In
   subjects:
   - kind: ...
      name: ...
   rule:
      apiGroups: ...

```

Reply:
`operator: NotIn` matches everything but what's listed --- but that's too much.  The usual use case is to carve exceptions out of a larger but not universal set.  In other words, "in list A but not in list B".

Reply:
> Another non-predicate approach would be to replace Subject rbacv1.Subject with Subjects []rbacv1.Subject and define a rbacv1.Subject value that means "all".

this approach sounds okay to me if we don't consider the "excluding" in the alpha stage. but in this way, i think we will have to expand the subject definition. and the "all" subject can be defined as:

```go
// const AllKind           = "All"
subjects:
- kind: "All"
```